runners/google-cloud-dataflow-java/worker/src/main/java/org/apache/beam/runners/dataflow/worker/WindmillStateInternals.java

```diff
@@ -2005,79 +2025,64 @@ public Iterable<Entry<K, V>> read() {
+                      KeyState keyState =
+                          keyStateMap.computeIfAbsent(structuralKey, k -> new KeyState(key));
+                      // Ignore any key from windmill that has been marked pending deletion, and any
+                      // key that is fully cached.
+                      if (keyState.existence == KeyExistence.KNOWN_NONEXISTENT
+                          || (keyState.existence == KeyExistence.KNOWN_EXIST
+                              && keyState.valuesCached)) return;
+                      // Or else cache contents from windmill.
+                      keyState.existence = KeyExistence.KNOWN_EXIST;
+                      keyState.values.extendWith(entry.getValue());
+                      // We can't set keyState.valuesCached to true here, because there may be more
+                      // paginated values that should not be filtered out in above if statement.
+                    } catch (IOException e) {
+                      throw new RuntimeException(e);
+                    }
                   });
               allKeysKnown = true;
               complete = true;
-              // Unload keys that are not known exist from cache, set valuesCached of all cached
-              // entries to true.
-              keyStateMap
-                  .entrySet()
-                  .removeIf(
-                      entry -> {
-                        KeyState keyState = entry.getValue();
-                        boolean shouldRemove =
-                            (keyState.existence == KeyExistence.KNOWN_NONEXISTENT
-                                    && !keyState.removedLocally)
-                                || keyState.existence == KeyExistence.UNKNOWN_EXISTENCE;
-                        keyState.valuesCached = !shouldRemove;
-                        return shouldRemove;
-                      });
               return Iterables.unmodifiableIterable(mergedCachedEntries());
             } else {
-              Iterable<Entry<K, V>> fromWindmill =
-                  Iterables.concat(
-                      Iterables.transform(
-                          Iterables.transform(
-                              filteredEntries,
-                              entry ->
-                                  new AbstractMap.SimpleEntry<>(
-                                      keyStateMap.get(entry.getKey()).originalKey,
-                                      entry.getValue())),
-                          entry ->
-                              Iterables.transform(
-                                  entry.getValue(),
-                                  v -> new AbstractMap.SimpleEntry<>(entry.getKey(), v))));
-              return Iterables.unmodifiableIterable(
-                  Iterables.concat(mergedCachedEntries(), fromWindmill));
+              MultimapIterables<K, V> cachedEntries = mergedCachedEntries();
+              entries.forEach(
```

Comment:
I think this is back to loading all the pages into the memory before iterating

Could you instead make it lazy by returning windmill non-cached data first (filtering cached or added to cached) then return the rest?

Iterables.concat(Iterables.transform( logic in entries.forEach), 
   some iterable that lazily calls mergedCacheEntries and iterates);
  
Maybe you can make some test for this by having the test have some iterable that would return GBs of data total which should be streamed through and not cause test to OOM?